### PR TITLE
[core] Do not consider X axis when constraining scale

### PR DESCRIPTION
**Pull request by brunoabinader**
Follow-up to https://github.com/mapbox/mapbox-gl-native/pull/12611, back then I did not realize that GL JS's behavior was to _not_ consider X axis when constraining scale. This PR fixes this edge case, and uses the newly added `basic-v9/z0-narrow-y` render test to verify parity with GL JS.

This fix uncovered an issue with heatmaps, where features outside the tile extent where considered in still rendering mode (@mourner you were right in https://github.com/mapbox/mapbox-gl-native/pull/12611#issuecomment-415004281 - it was indeed two points being rendered, one inside the tile extent, and the other outside).

### CODE REVIEW

**Comment by brunoabinader** [comment]
Noting that I ended up implementing support for fill-extrusion-vertical-gradient (https://github.com/mapbox/mapbox-gl-js/pull/6841), 99% of the code was auto-generated, I only needed to add a single line manually [here](https://github.com/mapbox/mapbox-gl-native/pull/13463/files#diff-059f83c328d3ade5d1e873e564b59d04R230) for it to work.

Fixes #13462.

File: platform/darwin/src/MGLFillExtrusionStyleLayer.h [hunk 1]
```diff
@@ -334,6 +334,27 @@ MGL_EXPORT
...
+ * Predefined functions, including mathematical and string operators
+ * Conditional expressions
+ * Variable assignments and references to assigned variables
+ * Step functions applied to the `$zoomLevel` variable
+ 
+ This property does not support applying interpolation functions to the
+ `$zoomLevel` variable or applying interpolation or step functions to feature
+ attributes.
+ */
+@property (nonatomic, null_resettable) NSExpression *fillExtrusionVerticalGradient;
```
Comment (1ec5) [inline platform/darwin/src/MGLFillExtrusionStyleLayer.h]: Paint and layout properties aren’t strongly typed in the iOS and macOS map SDK, so it’s important to conform to the Objective-C and Swift convention that Boolean-typed properties are named in a way that suggests a Boolean value. We’ll want to call this property `fillExtrusionHasVerticalGradient`. Override the property name as `fill-extrusion-has-vertical-gradient`:

https://github.com/mapbox/mapbox-gl-native/blob/ad03e42b2890fd430e99e543a40be42f09f47fba/platform/darwin/scripts/style-spec-cocoa-conventions-v8.json#L33-L36

The change will only affect the runtime styling API, not style JSON parsing. After adding the override, rerun `make darwin-style-code`.
